### PR TITLE
[Part 1] | TimeLock shoots itself

### PR DESCRIPTION
**Pull request by sudiksha27**
**Goals (and why)**:
Skeleton for TimeLock shooting itself if theres are signs of corruption.

**Implementation Description (bullets)**:

- Corruption detecting thread on local that kills itself and communicates to peers upon detecting TL corruption

- Add corruption pinger so TL can flag detection of corruption to its peers.

**Testing (What was existing testing like?  What have you done to improve it?)**:
N/A

**Concerns (what feedback would you like?)**:

- Does the design seem reasonable?

**Where should we start reviewing?**:
TimeLockLocalCorruptionDetector.java

**Priority (whenever / two weeks / yesterday)**:
Monday EOD would be great.

### CODE REVIEW

File: timelock-api/src/main/conjure/timelock-corruption.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+services:
+  TimeLockCorruptionPinger:
+    name: TimeLock Corruption Pinger
```
Comment (gmaretic) [inline timelock-api/src/main/conjure/timelock-corruption.yml:3]: Maybe TimeLock Corruption Notifier?

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/TimeLockCorruptionHealthCheck.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+public class TimeLockCorruptionHealthCheck {
+    private final List<CorruptionDetector> corruptionDetectors;
+
+    public TimeLockCorruptionHealthCheck(List<CorruptionDetector> corruptionDetectors) {
+        this.corruptionDetectors = corruptionDetectors;
+    }
+
+    public boolean isHealthy() {
+        return !(corruptionDetectors.stream().anyMatch(detector -> detector.hasDetectedCorruption()));
```
Comment (gmaretic) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/TimeLockCorruptionHealthCheck.java:29]:
```suggestion
        return corruptionDetectors.stream().noneMatch(CorruptionDetector::hasDetectedCorruption);
```

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/TimeLockCorruptionStatus.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.corruption;
+
+public enum TimeLockCorruptionStatus {
+    HEALTHY(false),
+    CORRUPTION(true);
+
+    private boolean hasCorruption;
```
Comment (gmaretic) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/TimeLockCorruptionStatus.java:23]:
```suggestion
    private final boolean hasCorruption;
```

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/TimeLockLocalCorruptionDetector.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    private static final AuthHeader AUTH_HEADER = AuthHeader.valueOf("Bearer omitted");
+    private static final Duration TIMELOCK_CORRUPTION_ANALYSIS_INTERVAL = Duration.ofMinutes(5);
+    private static final String CORRUPTION_DETECTOR_THREAD_PREFIX = "timelock-corruption-detector";
+
+    private final ScheduledExecutorService executor = PTExecutors.newSingleThreadScheduledExecutor(
+            new NamedThreadFactory(CORRUPTION_DETECTOR_THREAD_PREFIX, true));
+    private final List<TimeLockCorruptionPinger> corruptionPingers;
+
+    private TimeLockCorruptionStatus localCorruptionState = TimeLockCorruptionStatus.HEALTHY;
```
Comment (gmaretic) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/TimeLockLocalCorruptionDetector.java:44]:
```suggestion
    private volatile TimeLockCorruptionStatus localCorruptionState = TimeLockCorruptionStatus.HEALTHY;
```

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/TimeLockLocalCorruptionDetector.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        return timeLockLocalCorruptionDetector;
+    }
+
+    private TimeLockLocalCorruptionDetector(List<TimeLockCorruptionPinger> corruptionPingers) {
+        this.corruptionPingers = corruptionPingers;
+    }
+
+    private void scheduleWithFixedDelay() {
+        executor.scheduleWithFixedDelay(() -> {
+                    if (detectedSignsOfCorruption()) {
```
Comment (gmaretic) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/TimeLockLocalCorruptionDetector.java:61]: Depends on what this task exactly does, but would be nice if we could avoid checking every 5 minutes and detect corruption as part of normal operation

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/TimeLockLocalCorruptionDetector.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+    private void localDetectedCorruption() {
+        localCorruptionState = TimeLockCorruptionStatus.CORRUPTION;
+    }
+
+    private void reportCorruptionToRemote(TimeLockCorruptionPinger pinger) {
+        try {
+            pinger.corruptionDetected(AUTH_HEADER);
+        } catch (Exception e) {
+            log.warn("Failed to report corruption to remote.", e);
```
Comment (gmaretic) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/TimeLockLocalCorruptionDetector.java:83]: We will retry in 5 minutes I guess, but maybe we should retry sooner here

**Review by gmaretic — COMMENTED** [review]
Generally makes sense, just a few small things. Avoiding the task would be nice, but I imagine we can't do that -- it would still be good to proactively kill timelock if we can detect it before the task runs

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/LocalCorruptionHandler.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            List<TimeLockCorruptionNotifier> corruptionNotifiers) {
+        this.corruptionNotifiers = corruptionNotifiers;
+        this.executorService = PTExecutors.newFixedThreadPool(corruptionNotifiers.size());
+        this.corruptionNotifyRetryer = new Retryer<>(StopStrategies.stopAfterAttempt(5),
+                WaitStrategies.fixedWait(200, TimeUnit.MILLISECONDS),
+                attempt -> !attempt.hasResult());
+    }
+
+    public void notifyRemoteServersOfCorruption() {
+        corruptionNotifiers.forEach(this::submitTask);
```
Comment (gmaretic) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/LocalCorruptionHandler.java:56]:
```suggestion
        corruptionNotifiers.forEach(this::reportCorruptionToRemote);
```

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/LocalCorruptionHandler.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+    private static final AuthHeader AUTH_HEADER = AuthHeader.valueOf("Bearer omitted");
+    private final List<TimeLockCorruptionNotifier> corruptionNotifiers;
+    private final Retryer<Void> corruptionNotifyRetryer;
+    private final ExecutorService executorService;
+
+    public LocalCorruptionHandler(
+            List<TimeLockCorruptionNotifier> corruptionNotifiers) {
+        this.corruptionNotifiers = corruptionNotifiers;
+        this.executorService = PTExecutors.newFixedThreadPool(corruptionNotifiers.size());
```
Comment (gmaretic) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/LocalCorruptionHandler.java:49]: This is a bit too resource intensive now, as we are always creating this executor service, even when no corruption happened (also you forgot to shut it down eventually).
Also there is no need to do this asynchronously at this point really, so just run it synchronously

**Comment by gmaretic** [comment]
I see that you added the parallelism later, but it's not really needed so I'd prefer not to get too fancy :)

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/LocalCorruptionHandler.java (hunk 8 above)
Comment (sudiksha27) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/LocalCorruptionHandler.java]: Ah right, that makes sense.

**Review by gmaretic — APPROVED** [review]
LGTM

**Pull request merged**